### PR TITLE
More than fixes😱

### PR DESCRIPTION
**Pull request by IRacle1**
- ~Rooms init using patch(in my plans use patches to init all map stucts)~ reverted, old system was better
- Using equality comparer for components can be helpful
- reworked `SpawningItem` event, added `ShouldInitiallySpawn` and `TriggerDoor` properties, now you can make delayed items spawn initially, or inversely.
- fixed delayed pickup adding/removing.
- fixed IsSpawned property when delayed pickup spawns
can someone test plz😖😖😖

### CODE REVIEW

**Comment by IRacle1** [comment]
I tested that, but im not sure that absolutely everything works

File: Exiled.API/Features/ComponentsEqualityComparer.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        public bool Equals(Component x, Component y)
+        {
+            return x == y;
+        }
+
+        /// <inheritdoc/>
+        public int GetHashCode(Component obj)
+        {
+            return obj?.GetHashCode() ?? 0;
+        }
```
Comment (iopietro) [inline Exiled.API/Features/ComponentsEqualityComparer.cs:29]:
```suggestion
        public int GetHashCode(Component obj) => obj == null ? 0 : obj.GetHashCode();
```
Unfortunately `?.` and `??` operators cannot be used on classes that inherit from `UnityEngine.Object`.

File: Exiled.API/Features/ComponentsEqualityComparer.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    /// <summary>
+    /// Represents the equality comparer for unity components.
+    /// </summary>
+    internal class ComponentsEqualityComparer : IEqualityComparer<Component>
+    {
+        /// <inheritdoc/>
+        public bool Equals(Component x, Component y)
+        {
+            return x == y;
+        }
```
Comment (iopietro) [inline Exiled.API/Features/ComponentsEqualityComparer.cs:23]:
```suggestion
        public bool Equals(Component x, Component y) => x == y;
```

File: Exiled.API/Features/Door.cs [hunk 3]
```diff
@@ -265,9 +265,9 @@ public Vector3 Scale
         /// </summary>
         /// <param name="doorVariant">The base-game <see cref="DoorVariant"/>.</param>
         /// <returns>A <see cref="Door"/> wrapper object.</returns>
-        public static Door Get(DoorVariant doorVariant) => DoorVariantToDoor.ContainsKey(doorVariant)
+        public static Door Get(DoorVariant doorVariant) => doorVariant != null ? (DoorVariantToDoor.ContainsKey(doorVariant)
             ? DoorVariantToDoor[doorVariant]
```
Comment (iopietro) [inline Exiled.API/Features/Door.cs:269]:
```suggestion
        public static Door Get(DoorVariant doorVariant) => doorVariant != null ? (DoorVariantToDoor.TryGetValue(doorVariant, out Door door)
            ? door
```

File: Exiled.API/Features/Pickups/Pickup.cs [hunk 4]
```diff
@@ -428,10 +428,13 @@ public void UnSpawn()
...
         /// <summary>
-        /// Destroys the pickup.
+        /// Destroys the already spawned pickup.
         /// </summary>
         /// <seealso cref="UnSpawn"/>
-        public void Destroy() => Base.DestroySelf();
+        public void Destroy()
+        {
+            Base.DestroySelf();
+        }
```
Comment (iopietro) [inline Exiled.API/Features/Pickups/Pickup.cs:437]:
```suggestion
        public void Destroy() => Base.DestroySelf();
```

File: Exiled.Events/EventArgs/Map/SpawningItemEventArgs.cs [hunk 5]
```diff
@@ -7,38 +7,58 @@
...
         }
 
         /// <summary>
-        ///     Gets or sets a value indicating whether or not the item can be spawned.
+        ///     Gets a value indicating the pickup being spawned.
+        /// </summary>
+        public Pickup Pickup { get; }
+
+        /// <summary>
+        ///     Gets or sets a value indicating whether or not the item will be initially spawn.
```
Comment (iopietro) [inline Exiled.Events/EventArgs/Map/SpawningItemEventArgs.cs:46]:
```suggestion
        ///     Gets or sets a value indicating whether or not the item will be initially spawned.
```

File: Exiled.Events/Patches/Generic/PickupControlPatch.cs [hunk 6]
```diff
@@ -68,4 +67,32 @@ internal static class CreatePickupPatch
...
+    /// <summary>
+    /// Patches <see cref="ItemDistributor.SpawnPickup"/> to control <see cref="Pickup.IsSpawned"/> property for delayed spawned pickup.
+    /// </summary>
+    [HarmonyPatch(typeof(ItemDistributor), nameof(ItemDistributor.SpawnPickup))]
+    internal static class TriggerPickupControlPatch
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(
+            IEnumerable<CodeInstruction> instructions,
+            ILGenerator generator)
+        {
```
Comment (iopietro) [inline Exiled.Events/Patches/Generic/PickupControlPatch.cs:80]:
```suggestion
        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
        {
```

File: Exiled.Events/Patches/Generic/RoomListAdd.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    /// </summary>
+    [HarmonyPatch(typeof(RoomIdentifier), nameof(RoomIdentifier.Awake))]
+    internal class RoomListAdd
+    {
+        private static void Prefix(RoomIdentifier __instance)
+        {
+            Room.RoomIdentifierToRoom.Add(__instance, Room.CreateComponent(__instance.gameObject));
+        }
+
+        // idk, it throws me real time error
```
Comment (iopietro) [inline Exiled.Events/Patches/Generic/RoomListAdd.cs:35]: If the transpiler doesn't work, a prefix is fine for now, don't delete the comments.

**Review by iopietro — CHANGES_REQUESTED** [review]

**Pull request merged**